File: frontend/src/types/bicycle.ts
import { ProductDetail, ProductInfo } from './products';

// Type definitions for Bicycle Detail
export interface BicycleDetailProps {
  // Optional interface for bicycle details
  bicycleDetailInterface?: BicycleDetailInterface;
}

export interface BicycleDetailInterface {
  // Unique identifier for the bicycle
  id: number;
  // Name of the bicycle
  name: string;
  // Price of the bicycle
  price: number;
  // Indicates if the bicycle is sold
  sold: boolean;
  // Detailed information about the product
  productDetail: ProductDetail[];
}

export interface Bicycle {
  // Unique identifier for the bicycle
  id: number;
  // Name of the bicycle
  name: string;
  // Price of the bicycle
  price: number;
  // Indicates if the bicycle is sold
  sold: boolean;
  // Information about the product
  productInfo: ProductInfo[];
}

export interface BicyclePart {
  // Unique identifier for the bicycle part
  id: number;
  // Identifier for the bicycle this part belongs to
  bicycleId: number;
  // Identifier for the product this part is associated with
  productId: number;
  // Identifier for the product option
  productOptionId: number;
}

export interface UserCartData {
  // Unique identifier for the user
  user_id: number;
  // List of bikes in the user's cart
  bikes: { bike: Bicycle }[];
}

export interface ListBicyclePageState {
  // List of bicycles on the page
  bicycles: Bicycle[];
  // The currently selected bicycle, if any
  selectedBike: BicycleDetailInterface | null;
  // Indicates if the page is loading
  loading: boolean;
  // Error message, if any
  error: string | null;
}

// Props for the BicycleList component
export interface BicycleListProps {
  // List of bicycles to display
  bicycles: Bicycle[];
  // Function to handle the selection of a bicycle
  onSelectBike: (id: number) => void;
  // Indicates if the user is an admin
  isAdmin: boolean;
}

File: backend/app/models/bicycle.py
# app/models/bicycle.py
from .. import db

class Bicycle(db.Model):
    __tablename__ = 'bicycle'
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    name = db.Column(db.String(100), nullable=False)
    sold = db.Column(db.Boolean, default=False)
    price = db.Column(db.Float, nullable=False)

    def __init__(self, name, sold=False, price=0.0):
        self.name = name
        self.sold = sold
        self.price = price

    def to_dict(self):
        return {
            'id': self.id,
            'name': self.name,
            'sold': self.sold,
            'price': self.price
        }

class BicyclePart(db.Model):
    __tablename__ = 'bicycle_part'
    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    bicycle_id = db.Column(db.Integer, db.ForeignKey('bicycle.id'), nullable=False)
    product_id = db.Column(db.Integer, nullable=False)
    product_option_id = db.Column(db.Integer, nullable=False)

    def __init__(self, bicycle_id, product_id, product_option_id):
        self.bicycle_id = bicycle_id
        self.product_id = product_id
        self.product_option_id = product_option_id

    def to_dict(self):
        return {
            'id': self.id,
            'bicycle_id': self.bicycle_id,
            'product_id': self.product_id,
            'product_option_id': self.product_option_id
        }
    
